testcases/kernel/syscalls/fanotify/fanotify09.c

```diff
@@ -0,0 +1,236 @@
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <string.h>
+#include <sys/mount.h>
+#include <sys/syscall.h>
+#include "tst_test.h"
+#include "fanotify.h"
+
+#if defined(HAVE_SYS_FANOTIFY_H)
+#include <sys/fanotify.h>
+
+#define EVENT_MAX 1024
+/* size of the event structure, not counting name */
+#define EVENT_SIZE  (sizeof (struct fanotify_event_metadata))
+/* reasonable guess as to size of 1024 events */
+#define EVENT_BUF_LEN        (EVENT_MAX * EVENT_SIZE)
+
+static void setup(void);
+static void cleanup(void);
+
+#define NUM_GROUPS 3
+
+#define BUF_SIZE 256
+static char fname[BUF_SIZE];
+static int fd_notify[NUM_GROUPS];
+
+static char event_buf[EVENT_BUF_LEN];
+
+#define MOUNT_NAME "mntpoint"
+static int mount_created;
+
+static void create_fanotify_groups(void)
+{
+	unsigned int i, onchild;
+	int ret;
+
+	for (i = 0; i < NUM_GROUPS; i++) {
+		fd_notify[i] = SAFE_FANOTIFY_INIT(FAN_CLASS_NOTIF |
+						FAN_NONBLOCK,
+						O_RDONLY);
+
+		/* Add mount mark for each group without MODIFY event */
+		ret = fanotify_mark(fd_notify[i],
+				    FAN_MARK_ADD | FAN_MARK_MOUNT,
+				    FAN_CLOSE_NOWRITE,
+				    AT_FDCWD, ".");
+		if (ret < 0) {
+			tst_brk(TBROK | TERRNO,
+				 "fanotify_mark(%d, FAN_MARK_ADD | "
+				 "FAN_MARK_MOUNT, FAN_MODIFY, AT_FDCWD,"
+				 " '.') failed", fd_notify[i]);
+		}
+		/*
+		 * Add inode mark on parent for each group with MODIFY
+		 * event, but only one group requests events on child.
+		 * The one mark with FAN_EVENT_ON_CHILD is needed for
+		 * setting the DCACHE_FSNOTIFY_PARENT_WATCHED dentry
+		 * flag.
+		 */
+		onchild = (i == 0) ? FAN_EVENT_ON_CHILD : 0;
+		ret = fanotify_mark(fd_notify[i],
+				    FAN_MARK_ADD,
+				    FAN_MODIFY | onchild, AT_FDCWD, ".");
+		if (ret < 0) {
+			tst_brk(TBROK | TERRNO,
+				 "fanotify_mark(%d, FAN_MARK_ADD, "
+				 "FAN_MODIFY%s, AT_FDCWD, '.') failed",
+				 fd_notify[i],
+				 onchild ? " | FAN_EVENT_ON_CHILD" : "");
+		}
+	}
+}
+
+static void cleanup_fanotify_groups(void)
+{
+	unsigned int i;
+
+	for (i = 0; i < NUM_GROUPS; i++) {
+		if (fd_notify[i] && fd_notify[i] != -1) {
+			if (close(fd_notify[i]) == -1)
+				tst_res(TWARN, "close(%d) failed",
+					 fd_notify[i]);
+			fd_notify[i] = 0;
+		}
+	}
+}
+
+static void verify_event(int group, struct fanotify_event_metadata *event)
+{
+	if (event->mask != FAN_MODIFY) {
+		tst_res(TFAIL, "group %d get event: mask %llx (expected %llx) "
+			 "pid=%u fd=%u", group, (unsigned long long)event->mask,
+			 (unsigned long long)FAN_MODIFY,
+			 (unsigned)event->pid, event->fd);
+	} else if (event->pid != getpid()) {
+		tst_res(TFAIL, "group %d get event: mask %llx pid=%u "
+			 "(expected %u) fd=%u", group,
+			 (unsigned long long)event->mask, (unsigned)event->pid,
+			 (unsigned)getpid(), event->fd);
+	} else {
+		tst_res(TPASS, "group %d get event: mask %llx pid=%u fd=%u",
+			 group, (unsigned long long)event->mask,
+			 (unsigned)event->pid, event->fd);
+	}
+}
```

Comment:
We tend to write a bit more readable code with return inside of the if () body:

```
if (foo) {
         tst_res(TFAIL, ...);
         return;
}

if (bar) {
         tst_res(TFAIL, ...);
         return;
}

tst_res(TPASS, ...);
```

But that is very minor.

Reply:
Oh I addressed this comment on another part of the test, but not here.
Let me know if this needs changing.
Otherwise, rebased and addressed rest of the comments.

Reply:
I personally think that return makes the code flow easier to follow, but that is a matter of aesthetic and very minor...